Signal/src/util/OWSBackupExportJob.m

```diff
@@ -3,64 +3,226 @@
 //
 
 #import "OWSBackupExportJob.h"
+#import "OWSBackupIO.h"
 #import "OWSDatabaseMigration.h"
+#import "OWSSignalServiceProtos.pb.h"
 #import "Signal-Swift.h"
 #import <SignalServiceKit/NSData+Base64.h>
 #import <SignalServiceKit/NSDate+OWS.h>
 #import <SignalServiceKit/OWSBackgroundTask.h>
-#import <SignalServiceKit/OWSBackupStorage.h>
 #import <SignalServiceKit/OWSError.h>
 #import <SignalServiceKit/OWSFileSystem.h>
 #import <SignalServiceKit/TSAttachment.h>
 #import <SignalServiceKit/TSAttachmentStream.h>
 #import <SignalServiceKit/TSMessage.h>
 #import <SignalServiceKit/TSThread.h>
 #import <SignalServiceKit/Threading.h>
-#import <SignalServiceKit/YapDatabaseConnection+OWS.h>
-#import <YapDatabase/YapDatabase.h>
 
 NS_ASSUME_NONNULL_BEGIN
 
-NSString *const kOWSBackup_ExportDatabaseKeySpec = @"kOWSBackup_ExportDatabaseKeySpec";
+@class OWSAttachmentExport;
+
+@interface OWSBackupExportItem : NSObject
+
+@property (nonatomic) OWSBackupEncryptedItem *encryptedItem;
+
+@property (nonatomic) NSString *recordName;
+
+// This property is optional and represents the location of this
+// item relative to the root directory for items of this type.
+//@property (nonatomic, nullable) NSString *fileRelativePath;
+
+// This property is optional and is only used for attachments.
+@property (nonatomic, nullable) OWSAttachmentExport *attachmentExport;
+
+// This property is optional.
+@property (nonatomic, nullable) NSNumber *uncompressedDataLength;
+
+- (instancetype)init NS_UNAVAILABLE;
+
+@end
+
+#pragma mark -
+
+@implementation OWSBackupExportItem
+
+- (instancetype)initWithEncryptedItem:(OWSBackupEncryptedItem *)encryptedItem
+{
+    if (!(self = [super init])) {
+        return self;
+    }
+
+    OWSAssert(encryptedItem);
+
+    self.encryptedItem = encryptedItem;
+
+    return self;
+}
+
+@end
+
+#pragma mark -
+
+@interface OWSDBExportStream : NSObject
+
+@property (nonatomic) OWSBackupIO *backupIO;
+
+@property (nonatomic) NSMutableArray<OWSBackupExportItem *> *exportItems;
+
+@property (nonatomic, nullable) OWSSignalServiceProtosBackupSnapshotBuilder *backupSnapshotBuilder;
+
+@property (nonatomic) NSUInteger cachedItemCount;
+
+@property (nonatomic) NSUInteger totalItemCount;
+
+- (instancetype)init NS_UNAVAILABLE;
+
+@end
+
+#pragma mark -
+
+@implementation OWSDBExportStream
+
+- (instancetype)initWithBackupIO:(OWSBackupIO *)backupIO
+{
+    if (!(self = [super init])) {
+        return self;
+    }
+
+    OWSAssert(backupIO);
+
+    self.exportItems = [NSMutableArray new];
+    self.backupIO = backupIO;
+
+    return self;
+}
+
+- (BOOL)writeObject:(TSYapDatabaseObject *)object
+         entityType:(OWSSignalServiceProtosBackupSnapshotBackupEntityType)entityType
```

Comment:
It isn't strictly necessary to capture the entity type (the importer doesn't use this state), but I think it'll be helpful to have around to future-proof this work, help with debugging issue, etc.